### PR TITLE
Increased password requirements

### LINKED ISSUE #66: [Vulnerability] Stricter password requirements
* Enforce passwords with minimum 8 characters, at least 1 number,  at least 1 capital letter, and at least 1 symbol character

### LINKED ISSUE #76: Repeat password during registration for confirmation

### PR DESCRIPTION
**Pull request by ConorHK**
# Notes
* Closes #66 
* Closes #76 
## New requirements
*  8 - 64 characters
* Special character
* Uppercase character
* Lowercase character
* Number
password_fail
password_success
password_no_match

### CODE REVIEW

File: backend/src/main/java/com/vax/warden/validation/PasswordConstraintValidator.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                                new LengthRule(8, 64),
+                                new CharacterRule(EnglishCharacterData.UpperCase, 1),
+                                new CharacterRule(EnglishCharacterData.LowerCase, 1),
+                                new CharacterRule(EnglishCharacterData.Digit, 1),
+                                new CharacterRule(EnglishCharacterData.Special, 1)));
+        RuleResult result = validator.validate(new PasswordData(password));
+        if (result.isValid()) {
+            return true;
+        }
+        return false;
```
Comment (rajitbanerjee) [inline backend/src/main/java/com/vax/warden/validation/PasswordConstraintValidator.java:32]: :nit: verbosity
```suggestion
        return result.isValid();
```

File: backend/src/main/java/com/vax/warden/model/User.java [hunk 2]
```diff
@@ -69,6 +70,7 @@ public class User implements Serializable {
     @Enumerated(EnumType.STRING)
     private Gender gender;
 
+    @ValidPassword(message = "Password: Invalid password")
```
Comment (rajitbanerjee) [inline backend/src/main/java/com/vax/warden/model/User.java:73]: :nit:
```suggestion
    @ValidPassword(message = "Password: Password is not strong enough!")
```

File: frontend/src/pages/Registration.tsx [hunk 3]
```diff
@@ -127,13 +129,25 @@ export const Registration: React.FC = (): JSX.Element => {
             <FormControl isRequired marginTop={6}>
```
Comment (rajitbanerjee) [inline frontend/src/pages/Registration.tsx:129]: Need to show invalid password message from backend on Submit attempt
```suggestion
            <FormControl isRequired marginTop={6} isInvalid={formatUserDetailsKey.password in errorMap}>
```

File: frontend/src/pages/Registration.tsx [hunk 4]
```diff
@@ -127,13 +129,25 @@ export const Registration: React.FC = (): JSX.Element => {
...
+                  size="md"
+                  onChange={(event) => setPassword(event.target.value)}
+                />
                 <InputRightElement width="4.5rem">
                   <Button h="1.75rem" size="sm" onClick={handlePasswordShow} variant="ghost">
                     {showPassword ? <AiFillEyeInvisible /> : <AiFillEye />}
                   </Button>
                 </InputRightElement>
               </InputGroup>
+              <PasswordChecklist
```
Comment (rajitbanerjee) [inline frontend/src/pages/Registration.tsx:145]:
```suggestion
              </InputGroup>
              {getUserDetailsErrorMessage(errorMap, "password")}
              <PasswordChecklist
```
TODO: spacing between error message and password constraints list

**Review by rajitbanerjee — CHANGES_REQUESTED** [review]
Frontend nicely handled with minimal code!

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+pom.xml.releaseBackup
+pom.xml.versionsBackup
+pom.xml.next
+release.properties
+dependency-reduced-pom.xml
+buildNumber.properties
+.mvn/timing.properties
+# https://github.com/takari/maven-wrapper#usage-without-binary-jar
+.mvn/wrapper/maven-wrapper.jar
+
```
[the remark this answers is not in the document]
Comment (ConorHK) [reply .gitignore]: Merged into one in source

**Pull request merged**